ballerina/soap_utils.bal

```diff
@@ -1,106 +1,198 @@
+        secureSocket: config.httpConfig?.secureSocket,
+        circuitBreaker: config.httpConfig?.circuitBreaker
+    };
+}
+
+public function applySecurityPolicies(wssec:InboundSecurityConfig|wssec:InboundSecurityConfig[] inboundSecurity,
+                                      xml envelope) returns xml|wssec:Error {
+    wssec:InboundSecurityConfig|wssec:InboundSecurityConfig[] securityPolicy = inboundSecurity;
+    xml securedEnvelope;
+    if securityPolicy is wssec:InboundSecurityConfig {
+        if securityPolicy is wssec:TimestampTokenConfig {
+            securedEnvelope = check wssec:applyTimestampToken(envelope, securityPolicy);
+        } else if securityPolicy is wssec:UsernameTokenConfig {
+            securedEnvelope = check wssec:applyUsernameToken(envelope, securityPolicy);
+        } else if securityPolicy is wssec:SymmetricBindingConfig {
+            securedEnvelope = check wssec:applySymmetricBinding(envelope, securityPolicy);
+        } else if securityPolicy is wssec:AsymmetricBindingConfig {
+            securedEnvelope = check wssec:applyAsymmetricBinding(envelope, securityPolicy);
+        } else {
+            securedEnvelope = envelope;
+        }
+    } else {
+        foreach wssec:InboundSecurityConfig policy in securityPolicy {
+            securedEnvelope = check applySecurityPolicies(policy, envelope);
+        }
+    }
+    return securedEnvelope;
+}
+
+public function applyOutboundConfig(wssec:OutboundSecurityConfig outboundSecurity, xml envelope)
+    returns xml|Error {
+    xmlns "http://schemas.xmlsoap.org/soap/envelope/" as soap;
+    xml soapEnvelope = envelope;
+    do {
+        wssec:EncryptionAlgorithm? encryptionAlgorithm = outboundSecurity.decryptionAlgorithm;
+        if encryptionAlgorithm !is () {
+            crypto:PrivateKey|crypto:PublicKey? clientPrivateKey = outboundSecurity.decryptionKey;
+            if clientPrivateKey !is () {
+                byte[] encData = check wssec:getEncryptedData(soapEnvelope);
+                byte[] decryptDataResult = check wssec:decryptData(encData, encryptionAlgorithm, clientPrivateKey);
+                string decryptedBody = "<soap:Body >" + check string:fromBytes(decryptDataResult) + "</soap:Body>";
+                string decryptedEnv = regex:replace(soapEnvelope.toString(), string `<soap:Body .*>.*</soap:Body>`,
+                                                    decryptedBody);
+                soapEnvelope = check xml:fromString(decryptedEnv);
+            }
+        }
+        wssec:SignatureAlgorithm? signatureAlgorithm = outboundSecurity.signatureAlgorithm;
+        if signatureAlgorithm !is () {
+            crypto:PublicKey? serverPublicKey = outboundSecurity.verificationKey;
+            if serverPublicKey !is () {
+                byte[] signatureData = check wssec:getSignatureData(soapEnvelope);
+                boolean verify = check wssec:verifyData((soapEnvelope/<soap:Body>/*).toString().toBytes(),

```

Comment:
can't we directly return an error from verifyData method ?

Reply:
Fixed in [d7b7abf](https://github.com/ballerina-platform/module-ballerina-soap/pull/103/commits/d7b7abf894b23437dfb122a2303afdf16197a923)